build_docs.sh

```diff
@@ -0,0 +1,173 @@
+    DOCKER_RUN_ARGS+=('--tmpfs' '/run/nginx')
+    DOCKER_RUN_ARGS+=('--tmpfs' '/var/log/nginx')
+    DOCKER_RUN_ARGS+=('--tmpfs' '/var/lib/nginx/body')
+    DOCKER_RUN_ARGS+=('--tmpfs' '/var/lib/nginx/fastcgi')
+    DOCKER_RUN_ARGS+=('--tmpfs' '/var/lib/nginx/proxy')
+    DOCKER_RUN_ARGS+=('--tmpfs' '/var/lib/nginx/uwsgi')
+    DOCKER_RUN_ARGS+=('--tmpfs' '/var/lib/nginx/scgi')
+    echo "------------------------ WARNING ------------------------"
+    echo "$(basename "$0") can't open a browser. It'll start the web" \
+         "server but you must open the browser yourself."
+    echo "------------------------ WARNING ------------------------"
+    ;;
+  --out)
+    shift
+    DOCKER_RUN_ARGS+=('-v' "$(dirname "$(to_absolute_path $1)"):/out:delegated")
+    NEW_ARGS+=("/out/$(basename "$1")")
+    SAW_OUT=true
+    ;;
+  --reference)
+    shift
+    if [ ! -d "$1" ]; then
+      echo "Can't find $1"
+      exit 1
+    fi
+    DOCKER_RUN_ARGS+=('-v' "$(to_absolute_path $1):/reference:ro,cached")
+    NEW_ARGS+=("/reference")
+    ;;
+  --rely_on_ssh_auth)
+    if [ "$SSH_AUTH_SOCK" != "" ]; then
+      # If we have SSH auth share it into the container.
+      if [ "$(uname -s)" != Linux* ]; then
+        echo "------------------------ WARNING ------------------------"
+        echo "Attempting to share ssh auth but this is unlikely to work" \
+             "outside of linux."
+        echo "------------------------ WARNING ------------------------"
+      fi
+      DOCKER_RUN_ARGS+=('-v' "$(dirname $SSH_AUTH_SOCK):$(dirname $SSH_AUTH_SOCK):ro")
+      DOCKER_RUN_ARGS+=('-e' "SSH_AUTH_SOCK=$SSH_AUTH_SOCK")
+    fi
+    # Mount our known_hosts file into the VM so it won't ask about github
+    DOCKER_RUN_ARGS+=('-v' "$(to_absolute_path ~/.ssh/known_hosts):/tmp/.ssh/known_hosts:ro,cached")
+    ;;
+  --resource)
+    shift
+    if [ ! -d "$1" ]; then
+      echo "Can't find $1"
+      exit 1
+    fi
+    DOCKER_RUN_ARGS+=('-v' "$(to_absolute_path $1):/resource_$RESOURCE_COUNT:ro,cached")
+    NEW_ARGS+=("/resource_$RESOURCE_COUNT")
+    RESOURCE_COUNT+=1
+    ;;
+  *)
+    ;;
+  esac
+  shift
+done
+
+if [ "$SAW_OUT" = false ]; then
+  # If you don't specify --out then we dump the output into html_docs in the
+  # the current directory to keep backwards compatibility with build_docs.pl.
+  DOCKER_RUN_ARGS+=('-v' "$(dirname "$(to_absolute_path .)"):/out:delegated")
+  NEW_ARGS+=('--out' "/out/html_docs")
+fi
+
+tty -s
+if [ "0" == "$?" ]; then
+  # Emulate a terminal so things like ctrl-c work.
+  DOCKER_RUN_ARGS+=('-it')
+fi
+
+echo "Building the docker image that will build the docs. Expect this to" \
+    "take somewhere between a hundred milliseconds and five minutes."
+# Build the docker image from stdin so we don't try to pack up everything in
+# this directory. It is huge and we don't need any of it in the image because
+# we'll mount it into the image on startup.
+docker image build -t elastic/docs_build - < "$DIR/Dockerfile"
```

Comment:
Building the docker image right here feels good to me from a development background. It feels "like gradle but for native dependencies". I know the analogy isn't quite right, but it is pretty close in this case.

Reply:
I like that we could do things like `git bisect` and get the docker image specified by the Dockerfile in the repo right now. I'd be quite happy with a caching strategy that makes it so we don't have to build the image locally all the time, but this feels like a simple solution for now.